src/sentry/web/frontend/auth_login.py

```diff
@@ -67,11 +69,168 @@ def run_callbacks(self, request: Request):
+
+        request.session.set_test_cookie()
+
+        # Initiate login clears the pending session states so you don't end up at an odd location
+        initiate_login(request=request, next_url=next_uri)
+
+        # Single org mode -- send them to the org-specific handler
+        if settings.SENTRY_SINGLE_ORGANIZATION:
+            return self.redirect_to_org_specific_login()
+
+        session_expired = "session_expired" in request.COOKIES
+        if session_expired:
+            messages.add_message(
+                request=request, level=messages.WARNING, message=WARN_SESSION_EXPIRED
+            )
+
+        default_context = self.get_default_context(request=request)
+
+        if self.should_redirect_to_sso_login(request=request):
+            response = self.get_sso_redirect(request=request)
+        elif self.can_register(request=request):
+            response = self.get_registration_page(
+                request=request, context=default_context, **kwargs
+            )
+        else:
+            response = self.get_login_page(request=request, context=default_context, **kwargs)
+
+        if session_expired:
+            response.delete_cookie("session_expired")
+
+        return response
+
+    def get_next_uri(self, request: Request) -> str:
+        """
+        Returns the next URI a user should visit in their authentication flow.
+        """
+        next_uri_fallback = None
+        if request.session.get("_next") is not None:
+            next_uri_fallback = request.session.pop("_next")
+        return request.GET.get(REDIRECT_FIELD_NAME, next_uri_fallback)
+
+    def redirect_authenticated_user(self, request: Request, next_uri: str) -> HttpResponseRedirect:
+        """
+        If an authenticated user sends a GET request to AuthLoginView, we redirect them forwards in the auth process.
+        """
+        if is_valid_redirect(url=next_uri, allowed_hosts=(request.get_host(),)):
+            return self.redirect(url=next_uri)
+        return self.redirect_to_org(request=request)
+
+    def redirect_to_org_specific_login(self) -> HttpResponseRedirect:
+        """
+        If Sentry is in single org mode this redirects users to their specific handler.
+        """
+        org = Organization.get_default()
+        next_uri = reverse("sentry-auth-organization", args=[org.slug])
+        return HttpResponseRedirect(redirect_to=next_uri)
+
+    def should_redirect_to_sso_login(self, request: Request) -> bool:
+        """
+        Checks if a GET request to our login page should redirect to SSO login.
+        """
+        non_sso_urls = [
+            reverse("sentry-auth-organization", args=[request.subdomain]),
+            reverse("sentry-register"),
+        ]
+        return (
+            request.subdomain
+            and self.org_exists(request=request)
+            and request.path_info not in non_sso_urls
+        )
+
+    def get_sso_redirect(self, request: Request) -> HttpResponseRedirect:
```

Comment:
```suggestion
    def get_org_auth_login_redirect(self, request: Request) -> HttpResponseRedirect:
```
This isn't uniquely a SSO login redirect.

For example, both https://sentry.io/auth/login/ and https://sentry.io/auth/login/orgslug/ may be the same login template if `orgslug` hasn't set up any SSO provider. 

This just redirects to the login page specific for the organization (i.e. where the organization slug is fixed).